File: pkg/privateactionrunner/controlplane/rust/src/jwt.rs
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2026-present Datadog, Inc.

//! ES256 JWT generation, mirroring pkg/privateactionrunner/util/jwt.go.
//!
//! The Go code generates:
//!   header: {"alg":"ES256","typ":"JWT","cty":"JWT"}
//!   claims: {"orgId":<i64>, "runnerId":<str>, "iat":<unix>, "exp":<unix+60>}
//!
//! The private key is stored as a base64url-encoded JWK in datadog.yaml.
//! JWK format (EC P-256): {"kty":"EC","crv":"P-256","x":"...","y":"...","d":"..."}
//! The "d" field is the private scalar, base64url-encoded.

use std::time::{SystemTime, UNIX_EPOCH};

use anyhow::{Context, Result};
use base64::Engine;
use base64::engine::general_purpose::{URL_SAFE_NO_PAD, STANDARD};
use p256::ecdsa::{SigningKey, signature::Signer};
use serde_json::json;

/// Parses the base64url-encoded JWK and returns the P-256 signing key.
/// Called once at startup; the key is reused for every JWT.
pub fn load_signing_key(private_key_b64: &str) -> Result<SigningKey> {
    // The stored value is base64url (no padding) encoded JWK JSON.
    let jwk_bytes = URL_SAFE_NO_PAD
        .decode(private_key_b64)
        .or_else(|_| STANDARD.decode(private_key_b64))
        .context("failed to base64-decode private key")?;

    let jwk: serde_json::Value = serde_json::from_slice(&jwk_bytes)
        .context("failed to parse JWK JSON")?;

    let d_b64 = jwk["d"]
        .as_str()
        .ok_or_else(|| anyhow::anyhow!("JWK missing 'd' field"))?;

    let d_bytes = URL_SAFE_NO_PAD
        .decode(d_b64)
        .context("failed to base64url-decode JWK 'd' field")?;

    SigningKey::from_slice(&d_bytes)
        .context("failed to construct P-256 signing key from 'd' scalar")
}

/// Generates a signed ES256 JWT, mirroring GeneratePARJWT in jwt.go.
///
/// Header:  {"alg":"ES256","typ":"JWT","cty":"JWT"}
/// Claims:  {"orgId":<org_id>, "runnerId":<runner_id>, "iat":<now>, "exp":<now+60>}
/// Signature: ES256 over "<b64url_header>.<b64url_claims>" in IEEE P1363 format.
pub fn generate_jwt(org_id: i64, runner_id: &str, key: &SigningKey) -> Result<String> {
    let header = URL_SAFE_NO_PAD.encode(
        serde_json::to_string(&json!({"alg":"ES256","typ":"JWT","cty":"JWT"}))
            .context("failed to serialize JWT header")?
            .as_bytes(),
    );

    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .context("system clock before UNIX epoch")?
        .as_secs() as i64;

    let claims_json = serde_json::to_string(&json!({
        "orgId": org_id,
        "runnerId": runner_id,
        "iat": now,
        "exp": now + 60,
    }))
    .context("failed to serialize JWT claims")?;

    let claims = URL_SAFE_NO_PAD.encode(claims_json.as_bytes());

    let message = format!("{header}.{claims}");
    // p256::ecdsa::Signature::to_bytes() returns the IEEE P1363 representation
    // (r || s, 64 bytes) which is exactly what RFC 7518 §3.4 requires for ES256.
    let signature: p256::ecdsa::Signature = key.sign(message.as_bytes());
    let sig_bytes: Vec<u8> = signature.to_bytes().to_vec();
    let sig = URL_SAFE_NO_PAD.encode(&sig_bytes);

    Ok(format!("{message}.{sig}"))
}

File: pkg/privateactionrunner/shared/util/jwt.go
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2025-present Datadog, Inc.

package util

import (
	"crypto"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"maps"
	"time"

	"github.com/go-jose/go-jose/v4"
	"github.com/golang-jwt/jwt/v5"
)

func GenerateKeys() (*jose.JSONWebKey, *jose.JSONWebKey, error) {
	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to generate key pair: %w", err)
	}
	publicKey := &privateKey.PublicKey

	privateJwk, err := EcdsaToJWK(privateKey)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to encode private key: %w", err)
	}

	publicJwk, err := EcdsaToJWK(publicKey)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to encode public key: %w", err)
	}

	return privateJwk, publicJwk, nil
}

func Base64ToJWK(privateKey string) (jwk jose.JSONWebKey, err error) {
	decodedKeyBytes, err := base64.RawURLEncoding.DecodeString(privateKey)
	if err != nil {
		return jose.JSONWebKey{}, fmt.Errorf("error decoding private key: %+v", err)
	}
	if err = json.Unmarshal(decodedKeyBytes, &jwk); err != nil {
		return jose.JSONWebKey{}, fmt.Errorf("error converting private key to JWK: %+v", err)
	}
	return jwk, nil
}

func EcdsaToJWK(key any) (*jose.JSONWebKey, error) {
	// Check if the key is a ECDSA key.
	switch key.(type) {
	case *ecdsa.PrivateKey, *ecdsa.PublicKey:
	default:
		return nil, errors.New("unsupported key type")
	}

	// Encode the public key.
	newJwk := jose.JSONWebKey{
		Algorithm: "ES256",
		Key:       key,
		Use:       "sig",
	}

	// Compute the thumbprint of the public key.
	thumbprint, err := newJwk.Thumbprint(crypto.SHA256)
	if err != nil {
		return nil, err
	}

	// Set the key id.
	newJwk.KeyID = base64.RawURLEncoding.EncodeToString(thumbprint)

	return &newJwk, nil
}

func GeneratePARJWT(orgId int64, runnerId string, privateKey *ecdsa.PrivateKey, extraClaims map[string]any) (string, error) {
	claims := jwt.MapClaims{
		"orgId":    orgId,
		"runnerId": runnerId,
		"iat":      time.Now().Unix(), // this was added after version 1.7.0
		"exp":      time.Now().Add(time.Minute * 1).Unix(),
	}

	maps.Copy(claims, extraClaims)

	token := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
	token.Header["alg"] = "ES256"
	token.Header["cty"] = "JWT"

	signed, err := token.SignedString(privateKey)
	if err != nil {
		return "", fmt.Errorf("error signing JWT: %w", err)
	}

	return signed, nil
}

// JWKToPEM converts a JWK public key to PEM format
func JWKToPEM(pubJWK *jose.JSONWebKey) (string, error) {
	if !pubJWK.IsPublic() {
		return "", errors.New("error converting JWK to PEM: the key is not public")
	}

	pk, ok := pubJWK.Key.(*ecdsa.PublicKey)
	if !ok {
		return "", errors.New("error converting JWK to PEM: wrong underlying key type")
	}

	x509EncodedPub, err := x509.MarshalPKIXPublicKey(pk)
	if err != nil {
		return "", errors.New("error converting JWK to PEM: failed to marshal public key")
	}

	pemEncodedPub := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: x509EncodedPub})
	if pemEncodedPub == nil {
		return "", errors.New("error converting JWK to PEM: failed to encode public key to PEM format")
	}

	return string(pemEncodedPub), nil
}
